packages/react-router-dev/config/config.ts

```diff
@@ -515,9 +481,160 @@ export async function resolveReactRouterConfig({
     await preset.reactRouterConfigResolved?.({ reactRouterConfig });
   }
 
-  isFirstLoad = false;
+  return ok(reactRouterConfig);
+}
 
-  return reactRouterConfig;
+type ChokidarEventName = ChokidarEmitArgs[0];
+
+type ChangeHandler = (args: {
+  result: Result<ResolvedReactRouterConfig>;
+  configCodeUpdated: boolean;
+  configChanged: boolean;
+  routeConfigChanged: boolean;
+  path: string;
+  event: ChokidarEventName;
+}) => void;
+
+export type ConfigLoader = {
+  getConfig: () => Promise<Result<ResolvedReactRouterConfig>>;
+  onChange: (handler: ChangeHandler) => () => void;
+  close: () => Promise<void>;
+};
+
+export async function createConfigLoader({
+  rootDirectory: userRoot,
+  watch,
+}: {
+  watch: boolean;
+  rootDirectory?: string;
+}): Promise<ConfigLoader> {
+  let root = userRoot ?? process.env.REACT_ROUTER_ROOT ?? process.cwd();
+
+  let viteNodeContext = await ViteNode.createContext({
+    root,
+    mode: watch ? "development" : "production",
+    server: !watch ? { watch: null } : {},
+    ssr: {
+      external: ssrExternals,
+    },
+  });
+
+  let reactRouterConfigFile = findEntry(root, "react-router.config", {
+    absolute: true,
+  });
+
+  let getConfig = () =>
+    resolveConfig({ root, viteNodeContext, reactRouterConfigFile });
+
+  let appDirectory: string | undefined;
+
+  let initialConfigResult = await getConfig();
+  if (initialConfigResult.ok) {
```

Comment:
We should throw if first load failed. We can handle this better in the future, but for now this will do.